### PR TITLE
Ethereum "Get_Code" RPC call (CLI, Core, RestAPI + tests)

### PR DESCRIPTION
**Pull request by danilowhk**
Implemented "query-code" RPC call.

Implemented:
-CLI Commands + tests
-Beerus Core implementation + tests
-RestAPI implementation + tests

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: #76

# What is the new behavior?

-
-
-

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

# Other information

### CODE REVIEW

**Review by abdelhamidbakhta — APPROVED** [review]
Lgtm

File: beerus_cli/src/ethereum/mod.rs [hunk 1]
```diff
@@ -64,3 +64,20 @@ pub async fn query_block_number(beerus: BeerusLightClient) -> Result<CommandResp
...
+
+/// Query the code of a contract
+/// # Arguments
+/// * `beerus` - The Beerus light client.
+/// # Returns
+/// * `Result<CommandResponse>` - Vector of 256bits
+/// # Errors
+/// * If the block number query fails.
+pub async fn query_code(beerus: BeerusLightClient, address: String) -> Result<CommandResponse> {
+    let block = BlockTag::Latest;
```
Comment (LucasLvy) [inline beerus_cli/src/ethereum/mod.rs:76]: Can you make this configurable ? or open an issue to do it and add a `TODO`

File: beerus_cli/src/ethereum/mod.rs [hunk 2]
```diff
@@ -64,3 +64,20 @@ pub async fn query_block_number(beerus: BeerusLightClient) -> Result<CommandResp
...
+
+    let addr = Address::from_str(&address)?;
```
Comment (LucasLvy) [inline beerus_cli/src/ethereum/mod.rs:78]: Test this line plz (create a test that sends an invalid address to this func)

File: beerus_cli/src/model.rs [hunk 3]
```diff
@@ -120,6 +127,11 @@ impl Display for CommandResponse {
             CommandResponse::EthereumQueryBlockNumber(block_number) => {
                 write!(f, "{block_number}")
             }
+            // Print the code of a contract in 256bits vector
+            // Result looks like: [1,1,10,ff]
+            CommandResponse::EthereumQueryCode(code) => {
+                write!(f, "{code:?}")
```
Comment (LucasLvy) [inline beerus_cli/src/model.rs:133]: add opt to save the file (ex: `-o code.json`) or add issue and `TODO`

File: beerus_cli/src/runner.rs [hunk 4]
```diff
@@ -30,6 +30,9 @@ pub async fn run(beerus: BeerusLightClient, cli: Cli) -> Result<CommandResponse>
                 ethereum::query_nonce(beerus, address.to_string()).await
             }
             EthereumSubCommands::QueryBlockNumber {} => ethereum::query_block_number(beerus).await,
+            EthereumSubCommands::QueryCode { address } => {
+                ethereum::query_code(beerus, address.to_string()).await
```
Comment (LucasLvy) [inline beerus_cli/src/runner.rs:34]:
```suggestion
                ethereum::query_code(beerus, address.to_owned()).await
```

File: beerus_cli/tests/cli.rs [hunk 5]
```diff
@@ -127,6 +127,37 @@ mod test {
...
+            .expect_get_code()
+            .return_once(move |_, _| Ok(check_value));
+
+        let beerus = BeerusLightClient::new(
+            config,
+            Box::new(ethereum_lightclient),
+            Box::new(starknet_lightclient),
+        );
+
+        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_string();
```
Comment (LucasLvy) [inline beerus_cli/tests/cli.rs:147]:
```suggestion
        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_owned();
```

File: beerus_cli/tests/cli.rs [hunk 6]
```diff
@@ -243,6 +274,44 @@ mod test {
...
+            .expect_get_code()
+            .return_once(move |_, _| Err(eyre::eyre!("ethereum_lightclient_error")));
+
+        let beerus = BeerusLightClient::new(
+            config,
+            Box::new(ethereum_lightclient),
+            Box::new(starknet_lightclient),
+        );
+
+        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_string();
```
Comment (LucasLvy) [inline beerus_cli/tests/cli.rs:295]:
```suggestion
        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_owned();
```

File: beerus_cli/tests/cli.rs [hunk 7]
```diff
@@ -243,6 +274,44 @@ mod test {
...
+                command: EthereumSubCommands::QueryCode { address },
+            }),
+        };
+
+        // When
+        let result = runner::run(beerus, cli).await;
+
+        // Then
+        match result {
+            Err(e) => assert_eq!("ethereum_lightclient_error", e.to_string()),
```
Comment (LucasLvy) [inline beerus_cli/tests/cli.rs:310]:
```suggestion
            Err(e) => assert_eq!("ethereum_lightclient_error", e.to_owned()),
```

File: beerus_core/tests/beerus.rs [hunk 8]
```diff
@@ -219,6 +219,81 @@ mod tests {
...
+            .expect_get_code()
+            .return_once(move |_, _| Ok(expected_code));
+        let beerus = BeerusLightClient::new(
+            config.clone(),
+            Box::new(ethereum_lightclient_mock),
+            Box::new(starknet_lightclient_mock),
+        );
+
+        // Prepare variables
+        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_string();
```
Comment (LucasLvy) [inline beerus_core/tests/beerus.rs:242]:
```suggestion
        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_owned();
```

File: beerus_core/tests/beerus.rs [hunk 9]
```diff
@@ -219,6 +219,81 @@ mod tests {
...
+            .return_once(move |_, _| Err(eyre::eyre!("ethereum_lightclient_error")));
+
+        // When
+        let beerus = BeerusLightClient::new(
+            config.clone(),
+            Box::new(ethereum_lightclient_mock),
+            Box::new(starknet_lightclient_mock),
+        );
+
+        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_string();
```
Comment (LucasLvy) [inline beerus_core/tests/beerus.rs:281]:
```suggestion
        let address = "0xc24215226336d22238a20a72f8e489c005b44c4a".to_owned();
```

File: beerus_core/src/lightclient/ethereum/helios_lightclient.rs [hunk 10]
```diff
@@ -48,6 +48,14 @@ impl EthereumLightClient for HeliosLightClient {
     async fn get_block_number(&self) -> eyre::Result<u64> {
         self.helios_light_client.get_block_number().await
     }
+
+    async fn get_code(
+        &self,
+        address: &ethers::types::Address,
```
Comment (LucasLvy) [inline beerus_core/src/lightclient/ethereum/helios_lightclient.rs:54]:
```suggestion
        address: &Address,
```
import `ethers::types::Address`

File: beerus_core/src/lightclient/ethereum/helios_lightclient.rs [hunk 11]
```diff
@@ -48,6 +48,14 @@ impl EthereumLightClient for HeliosLightClient {
...
+        block: helios::types::BlockTag,
+    ) -> eyre::Result<Vec<u8>> {
```
Comment (LucasLvy) [inline beerus_core/src/lightclient/ethereum/helios_lightclient.rs:56]:
```suggestion
    ) -> Result<Vec<u8>> {
```
import `eyre::Result`

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: beerus_cli/src/ethereum/mod.rs (hunk 1 above)
Comment (danilowhk) [reply beerus_cli/src/ethereum/mod.rs:76]: Ok!

Comment (danilowhk) [reply beerus_cli/src/ethereum/mod.rs:77]: When u mean configurable would it be to receive it as an argument?

File: beerus_cli/src/model.rs (hunk 3 above)
Comment (danilowhk) [reply beerus_cli/src/model.rs:138]: The idea would be that we will have a bool or enum option to return a json file aswell? Would this option be taken as argument or just manually change?

File: beerus_cli/src/ethereum/mod.rs (hunk 2 above)
Comment (danilowhk) [reply beerus_cli/src/ethereum/mod.rs:79]: Implemented test on beerus_cli/tests/clir.rs line 674

File: beerus_cli/src/ethereum/mod.rs (hunk 1 above)
Comment (LucasLvy) [reply beerus_cli/src/ethereum/mod.rs:77]: yes

File: beerus_cli/src/model.rs (hunk 3 above)
Comment (LucasLvy) [reply beerus_cli/src/model.rs:137]: No it'd be an optional path if provided then the code is dumped into the file if not provided simply print it as you're doing now.

File: beerus_cli/tests/cli.rs [hunk 12]
```diff
@@ -56,6 +56,46 @@ mod test {
...
+            Box::new(ethereum_lightclient),
+            Box::new(starknet_lightclient),
+        );
+
+        // Mock the command line arguments.
+        let cli = Cli {
+            config: None,
+            command: Commands::Ethereum(EthereumCommands {
+                command: EthereumSubCommands::QueryBalance {
+                    address: "0xc24215226336d22238a20a72f8e489c005b44c4a".to_string(),
```
Comment (LucasLvy) [inline beerus_cli/tests/cli.rs:85]:
```suggestion
                    address: "0xc24215226336d22238a20a72f8e489c005b44c4a".to_owned(),
```

File: beerus_cli/tests/cli.rs [hunk 13]
```diff
@@ -93,6 +133,45 @@ mod test {
...
+            Box::new(ethereum_lightclient),
+            Box::new(starknet_lightclient),
+        );
+
+        // Mock the command line arguments.
+        let cli = Cli {
+            config: None,
+            command: Commands::Ethereum(EthereumCommands {
+                command: EthereumSubCommands::QueryNonce {
+                    address: "0xc24215226336d22238a20a72f8e489c005b44c4a".to_string(),
```
Comment (LucasLvy) [inline beerus_cli/tests/cli.rs:161]:
```suggestion
                    address: "0xc24215226336d22238a20a72f8e489c005b44c4a".to_owned(),
```

**Review by LucasLvy — APPROVED** [review]

**Comment by danilowhk** [comment]
I have rebased but still appears as "cannot be rebased due to conflicts" , should I close and reopen a new PR?

**Pull request closed without merging**